include/util_quda.h

```diff
@@ -7,15 +7,42 @@
 #include <tune_key.h>
 #include <malloc_quda.h>
 
+#ifndef QUDA_MUSTTAIL
+#ifdef __clang__
+#define QUDA_MUSTTAIL __attribute__((musttail))
```

Comment:
What does this attribute do with clang?  Is this is safe to include for all clang-based compilation targets, e.g., ROCm, clang-cuda, etc.?

Reply:
Here, https://clang.llvm.org/docs/AttributeReference.html#musttail. On the host, instead of preparing the stack pointer, argument passing, and `call` a function, the compiler should generate a `jmp`, reusing the existing stack frame.  So it's guaranteed that this recursive template expansion would not grow the stack even some how the functions are not inlined.  I'm not entirely sure what it would do on a target device.  I guess it may not matter if the recursive function is always inlined, or the compiler is smart enough.  We should probably check the assembly to be sure.